[language] implement Diem transactional test runner

vgao1996 opened the pull request:
This is an implementation of a Diem adapter based transactional test runner, built on top of @tnowacki's [framework](https://github.com/diem/diem/pull/8766).

## Overview
The contents of the PR are split into 3 commits:
- The first commit manually implements StructOpt for the generic command type so we don't have to mess with macros and write a split function
- The second commit moves some key generation logic from `language-e2e-tests` into a new crate called `diem-keygen`. This is also a binary you can run that will print you a pair of public and private Ed25519 keys. I'm expecting this to be used a lot by test writers.
- The third commit is the initial implementation of the Diem transactional test runner. 

Currently the Diem transactional test runner has only the basic capabilities, meaning that it doesn't allow you to set a lot of transaction parameters. However, the infrastructure to support these is already included and I'm expecting us to add these configuration options over time, while we migrate the tests.

## TODOs
- [x] Implement view command
- [x] Do not pull in FakeExecutor
-  ~~Fix some parsing bugs in the framework~~ (separate PR)
- [x] Handling of error paths
- [x] More tests
- [x] Documentation

tnowacki commented on language/diem-tools/diem-keygen/src/main.rs:
```diff
@@ -0,0 +1,15 @@
+    let mut keygen = KeyGen::from_os_rng();
+    let (privkey, pubkey) = keygen.generate_keypair();
+
+    println!("Private Key:");
```
?

vgao1996 replied:
Well, since the test harness does not currently generate keys for you, you'll have to do it manually a lot when writing tests. I made this executable to make that work slightly less tedious.

tnowacki commented on language/testing-infra/diem-transactional-test-runner/Cargo.toml:
```diff
@@ -0,0 +1,36 @@
+[package]
+name = "diem-transactional-test-runner"
```
do we need to put this in a separate dir for move/diem separation? Like, should it be in the language/ dir?

And minor nit, but maybe `diem-transactional-test-harness` not sure it really matters though

vgao1996 replied:
Eventually we'll definitely move this away current location, but I don't think now is the right time. I think we should at least come up with some rough guidelines on how we relocate diem crates in language before we start to tackle individual cases.

I'll change the name to `diem-transactional-test-harness` as it sounds slightly more accurate.

tnowacki commented on language/testing-infra/diem-transactional-test-runner/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,320 @@
+        ScriptFunction as TransactionScriptFunction,
+    },
+};
+use language_e2e_tests::executor::FakeExecutor;
```
Do we need to pull this in? I was imagining it would be easy enough to just talk to the diem adapter directly

tnowacki replied:
Looking at it a bit closer, it should be easier now than it was in the past to just talk to diem_vm.rs right?

vgao1996 replied:
Agreed we should get rid of it, but I'm currently depending on it to fetch some on-chain resources. Let me see what I can do. I'll try to get rid of it, but I have feeling that this will be a lot of work. If that's the case then I'll likely do it in a separate PR.

tnowacki replied:
I'm thinking it might actually be easier to start with a very basic invocation of `DiemVM` + `InMemoryStorage`. Reading or editing these stateful resources probably isn't needed in the first PR

vgao1996 replied:
> Reading or editing these stateful resources probably isn't needed in the first PR

It's needed because we need to fetch certain transaction parameters which are again, required to perform any transactions. If you don't fetch the default values from the chain, you'll have to require many more additional command arguments, which will be quite messy IMO.

tnowacki replied:
Ah gotcha

... I would still be curious to see one without pulling in FakeExecutor though :P Our goal is to kill language/e2e-testsuite as its all kind of crufty and out dated. My unsubstantiated and unresearched belief is that the new thing can be built a lot simpler

tnowacki commented on language/testing-infra/diem-transactional-test-runner/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,320 @@
+
+#[derive(StructOpt, Debug)]
+struct DiemPublishArgs {
+    #[structopt(short = "k", long = "private-key", parse(try_from_str = parse_ed25519_private_key))]
```
Do we need to support this? I'd rather see it automanaged. We are testing the DF, not key management :)

vgao1996 replied:
Well a private key is something you MUST provide in order to execute a signed transaction, just like the real blockchain. We can make the test adapter maintain a key mapping, similar to what we have in the old functional tests, but I thought we decided to favor explicitness and not do this based on our previous discussions?

tnowacki commented on language/testing-infra/diem-transactional-test-runner/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,320 @@
+            named_address_mapping.insert(name, addr);
+        }
+
+        // TODO: for now, we are using the FakeExecutor to keep things simple. This however,
```
maybe another good reason not to use FakeExecutor :)

tnowacki commented on language/testing-infra/diem-transactional-test-runner/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,320 @@
+        )?
+        .into_inner();
+
+        // TODO: handle vm status
```
Maybe make my status handler public, or utility-ify some things from the vm_test_harness

vgao1996 replied:
I'll look into those options

tnowacki reviewed (commented):
Looks like a decent start :) Mostly worried about the key management stuff and the presence of the FakeExecutor

vgao1996 commented:
@tnowacki I made the changes you requested
- Got rid of the reliance on `FakeExecutor`
- Renamed the crate to `diem-transactional-test-harness`

Also implemented some missing features
- View command
- Error handling

What I'm not going to include in this PR:
- Private key mapping + support for using named addresses as command arguments
- Configurable transaction parameters
- Fix to the command parser bug

tnowacki commented on language/testing-infra/diem-transactional-test-harness/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,395 @@
+    /// Obtain a Rust representation of the account resource from storage, which is used to derive
+    /// a few default transaction parameters.
+    fn fetch_account_resource(&self, signer_addr: &AccountAddress) -> Result<AccountResource> {
+        let account_access_path = AccessPath::resource_access_path(ResourceKey::new(
+            *signer_addr,
+            AccountResource::struct_tag(),
+        ));
```
These don't look they weren't that bad to implement. Glad we kicked `FakeExecutor`!

tnowacki commented on language/testing-infra/diem-transactional-test-harness/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,395 @@
+                    self.storage.add_write_set(output.write_set());
+                }
+                _ => {
+                    bail!("Failed to execute transaction. VMStatus: {}", status)
```
We might want to change the printing here slightly (I had to for the VMError stuff). but it is fine for now, and very easy to change

tnowacki commented on language/testing-infra/transactional-test-runner/src/tasks.rs:
```diff
@@ -236,68 +236,85 @@ pub struct ViewCommand {
+// a backward-incompatible way, they'll still need to offer the ability to extract arguments
+// from an `App` in one way or another, or otherwise `#[struct_opt(flatten)`, which is a public
+// API, will get broken.
+impl<ExtraInitArgs, ExtraPublishArgs, ExtraRunArgs, SubCommands> StructOpt
```
Fancy!

tnowacki commented on language/testing-infra/diem-transactional-test-harness/src/diem_test_harness.rs:
```diff
@@ -0,0 +1,395 @@
+///   - It executes transactions through DiemVM, instead of MoveVM directly
+struct DiemTestAdapter<'a> {
+    compiled_state: CompiledState<'a>,
+    storage: FakeDataStore,
```
No chance of using `InMemoryStorage` here? 
I can always play around with that if needed

vgao1996 replied:
One thing is for sure: we want a StateView-based storage backend since that is what the Diem adapter expects. Now it isn't entirely impossible to implement StateView for InMemoryStorage, but I'd like think little more about its interaction with access paths...

tnowacki reviewed (approved):
Awesome we were able to kill `KnownCommandFormat`. This looks really polished and good, even with some small things left to change in the future (like maybe ditching `FakeDataStore`)

vgao1996 commented:
/land
